File: ReorderList/ReorderList.cpp
// ReorderList.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>

class ListNode
{
public:
    int data;
    ListNode* next;
    ListNode(int data)
        : data(data)
        , next(nullptr)
    {}
};

ListNode* GetListFromInput(ListNode** ppHead, int count)
{
    if (count == 0)
        return nullptr;

    std::cout << "Enter " << count << " numbers of the list: ";
    ListNode* pTail = *ppHead;
    for (int i = 0; i < count; ++i)
    {
        int data = 0;
        std::cin >> data;
        if (*ppHead == nullptr)
        {
            *ppHead = new ListNode(data);
            pTail = *ppHead;
        }
        else
        {
            pTail->next = new ListNode(data);
            pTail = pTail->next;
        }
    }

    return pTail;
}

void PrintList(ListNode* pHead)
{
    if (pHead == nullptr)
        return;

    std::cout << pHead->data;
    pHead = pHead->next;

    while (pHead != nullptr)
    {
        std::cout << ", " << pHead->data;
        pHead = pHead->next;
    }
}

void FreeList(ListNode* pNode)
{
    if (pNode == nullptr)
        return;

    FreeList(pNode->next);
    delete pNode;
}

void SwitchNodes(ListNode** ppHead, ListNode* pTail)
{
    if (pTail == nullptr)
        return;     // only occurs when there is 1 element

    SwitchNodes(ppHead, pTail->next);

    if (*ppHead == nullptr)
        return;
    if (*ppHead == pTail || (*ppHead)->next == pTail)
    {
        pTail->next = nullptr;
        *ppHead = nullptr;
        return;
    }
    
    ListNode* pNextHead = (*ppHead)->next;
    (*ppHead)->next = pTail;
    pTail->next = pNextHead;
    *ppHead = pNextHead;
}

ListNode* ReorderList(ListNode* head)
{
    ListNode* pHead = head;
    SwitchNodes(&pHead, head);
    return head;
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of nodes in list: ";
        std::cin >> count;
        if (count <= 0)
            break;

        ListNode* pHead = nullptr;
        ListNode* pTail = GetListFromInput(&pHead, count);

        ReorderList(pHead);

        std::cout << "The re-ordered list: ";
        PrintList(pHead);
        std::cout << std::endl;

        FreeList(pHead);
    }
}

File: Go/ReorderList.go
package main

import "fmt"

type ListNode struct {
	data int
	next *ListNode
}

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetListFromArray(v []int) *ListNode {
	if len(v) == 0 {
		return nil
	}

	if len(v) == 1 {
		return &ListNode{v[0], nil}
	}

	pHead := &ListNode{v[0], nil}
	pCur := pHead
	for i := 1; i < len(v); i++ {
		pCur.next = &ListNode{v[i], nil}
		pCur = pCur.next
	}

	return pHead
}

func (pNode *ListNode) Print() {
	fmt.Print(pNode.data)
	if pNode.next != nil {
		fmt.Print(", ")
		pNode.next.Print()
	} else {
		fmt.Println("")
	}
}

func Reorder(pHead, pTail *ListNode) *ListNode {
	if pTail == nil {
		return pHead
	}

	pNode := Reorder(pHead, pTail.next)
	if pNode == nil {
		return nil
	}

	if pNode == pTail || pNode.next == pTail {
		pTail.next = nil
		return nil
	}

	pNextHead := pNode.next
	pNode.next = pTail
	pTail.next = pNextHead

	return pNextHead
}

func ReorderList(head *ListNode) *ListNode {
	Reorder(head, head.next)
	return head
}

func main() {
	pHead := &ListNode{1, &ListNode{2, &ListNode{3, &ListNode{4, &ListNode{5, &ListNode{6, &ListNode{7, &ListNode{8, &ListNode{9, nil}}}}}}}}}

	ReorderList(pHead)

	pHead.Print()
}
